### PR TITLE
fix(listview): update index after items removed

### LINKED ISSUE #5114: Calling `remove_items` or `pop` on a ListView does not update index or highlighting
It seems that calling `pop` (or `remove_items`, though that's not what I'm using for my use case) does not correctly update highlighting (presumably because calling `pop` does not update the index).

A small reproduction is as follows:

```python
from textual.app import App, ComposeResult
from textual.widgets import ListItem, ListView, Footer


class TestApp(App):
    BINDINGS = [
        ('a', 'add', 'Adds a blank item'),
        ('d', 'delete_highlighted', 'Deletes the highlighted cell'),
    ]

    _listview: ListView

    def compose(self) -> ComposeResult:
        self._listview = ListView(ListItem())

        yield Footer()
        yield self._listview

    def action_add(self):
        self._listview.append(ListItem())

    def action_delete_highlighted(self):
        index = self._listview.index
        self._listview.pop(index)


if __name__ == '__main__':
    TestApp().run()
```

Spawning a few items and trying to delete them does not set any highlighting, only manually updating the index will do that.

# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.83.0 |
| Rich    | 13.9.2 |

## Python

| Name           | Value                                       |
|----------------|---------------------------------------------|
| Version        | 3.12.1                                      |
| Implementation | CPython                                     |
| Compiler       | Clang 17.0.6                                |
| Executable     | /home/king/Projects/private/.venv/bin/python |

## Operating System

| Name    | Value                                                |
|---------|------------------------------------------------------|
| System  | Linux                                                |
| Release | 6.10.11-amd64                                        |
| Version | #1 SMP PREEMPT_DYNAMIC Debian 6.10.11-1 (2024-09-22) |

## Terminal

| Name                 | Value          |
|----------------------|----------------|
| Terminal Application | *Unknown*      |
| TERM                 | xterm-256color |
| COLORTERM            | truecolor      |
| FORCE_COLOR          | *Not set*      |
| NO_COLOR             | *Not set*      |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=223, height=76 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 223                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 76                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |

### PR DESCRIPTION
**Pull request by TomJGooding**
Fixes #5114. 

- The `pop` and `remove_items`  methods now return  `AwaitComplete` rather than `AwaitRemove` to account for the index update
- Previously trying to `pop()` from an empty list would raise a `NoMatches` error, this will now raise an `IndexError` for consistency.

**Please review the following checklist.**

- [ ] Docstrings on all new or modified functions / classes 
- [ ] Updated documentation
- [x] Updated CHANGELOG.md (where appropriate)

### CODE REVIEW

**Comment by TomJGooding** [comment]
I've now also updated the `remove_items` method, so hopefully I was on the right track!

**Comment by TomJGooding** [comment]
This now also reinstates `always_update` to the `index` reactive that was removed in recent changes. When items are removed, the `index` may not change but the highlighted item should be updated.

**Comment by willmcgugan** [comment]
I'd recommend against the `always_update`. Since it calls watchers you can get unintended side effects, such as re-sending messages.

Better to factor out the specific functionality you need and call it directly.

TBH I regret adding `always_update`.  I know we've recommended it in the past, but this will rightly confuse anything not intimately familiar with reactives:

```python
self.index = self.index
```

File: src/textual/widgets/_list_view.py
Comment (willmcgugan) [inline src/textual/widgets/_list_view.py:297]: Docstring and typing  por favor. Since its internal, and essentially private, the docstrings can be written for future Textual core-devs.

File: src/textual/widgets/_list_view.py
Comment (willmcgugan) [inline src/textual/widgets/_list_view.py:328]: docstrings please

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Just a request for docstrings...

File: src/textual/widgets/_list_view.py
Comment (TomJGooding) [reply src/textual/widgets/_list_view.py]: I've added docstrings in 9314cde, but feel free to tweak!

**Comment by willmcgugan** [comment]
Hi Tom, is this ready to go?

**Comment by darrenburns** [comment]
@willmcgugan Was going to merge, but it's blocked until you approve.

**Review by willmcgugan — APPROVED** [review]

**Pull request merged**
